core/templates/dev/head/collection_editor/CollectionEditorSavePublishDirective.js

```diff
@@ -0,0 +1,139 @@
+  return {
+    restrict: 'E',
+    scope: {
+      getCollection: '&collection'
+    },
+    templateUrl: 'inline/collection_editor_save_publish_directive',
+    controller: [
+      '$scope', '$modal', 'UndoRedoService', 'CollectionValidationService',
+      'CollectionRightsBackendApiService',
+      'WritableCollectionBackendApiService',
+      function(
+          $scope, $modal, UndoRedoService, CollectionValidationService,
+          CollectionRightsBackendApiService,
+          WritableCollectionBackendApiService) {
+        $scope.isPrivate = GLOBALS.isPrivate;
+        $scope.canUnpublish = GLOBALS.canUnpublish;
+        $scope.validationIssues = [];
+        $scope.isSaveInProgress = false;
+        var _collectionId = GLOBALS.collectionId;
+
+        var _validateCollection = function() {
+          if ($scope.isPrivate) {
+            $scope.validationIssues = (
+              CollectionValidationService.
+                findValidationIssuesForPrivateCollection(
+                  $scope.getCollection()));
+          } else {
+            $scope.validationIssues = (
+              CollectionValidationService.
+                findValidationIssuesForPublicCollection(
+                  $scope.getCollection()));
+          }
+        };
+
+        $scope.$on('collectionLoaded', function() {
+          _validateCollection();
+        });
+
+        UndoRedoService.setOnChangedCallback(_validateCollection);
+
+        $scope.getChangeListCount = function() {
+          return UndoRedoService.getChangeCount();
+        };
+
+        $scope.isCollectionSaveable = function() {
+          return (
+            $scope.getChangeListCount() > 0 &&
+            $scope.validationIssues.length === 0);
+        };
+
+        $scope.isCollectionPublishable = function() {
+          return (
+            $scope.isPrivate &&
+            $scope.getChangeListCount() === 0 &&
+            $scope.validationIssues.length === 0);
+        };
+
+        // An explicit save is needed to push all changes to the backend at
+        // once because some likely working states of the collection will cause
+        // validation errors when trying to incrementally save them.
+        $scope.saveCollection = function(commitMessage) {
+          // Don't attempt to save the collection if there are no changes
+          // pending.
+          if (!UndoRedoService.hasChanges()) {
+            return;
+          }
+          $scope.isSaveInProgress = true;
+          WritableCollectionBackendApiService.updateCollection(
+            $scope.getCollection().getId(),
+            $scope.getCollection().getVersion(),
+            commitMessage, UndoRedoService.getCommittableChangeList()).then(
+            function(collectionBackendObject) {
+              // TODO(sll): This needs to update the collection object in the
+              // CollectionEditor.js parent controller. It currently fails.
+              _updateCollection(collectionBackendObject);
```

Comment:
@BenHenning -- flagging this. We need a system for updating the $scope.collection in the frontend.